bgls/simulator_test.py

```diff
@@ -278,3 +280,60 @@ def test_remains_clifford():
     sim_results_stab = sim_act_on_stab.run(clifford_circuit, repetitions=100)
 
     assert sim_results == sim_results_stab
+
+
+def test_mps_results_match_state_vec():
+    """Test sampled bitstrings are same when using a matrix product state
+    simulator and a state vector simulator.
+    """
+    try:
+        import quimb.tensor as qtn
+        import cirq.contrib.quimb.mps_simulator
+    except ImportError:
+        return "quimb not installed, mps simulation not possible"
+
+    def cirq_mps_bitstring_probability(
+        mps: cirq.contrib.quimb.MPSState, bitstring: str
+    ) -> float:
+        """
+        Returns the probability of measuring the `bitstring` (|z⟩) in the
+        'cirq.contrib.quimb.MPSState' mps.
+        Args:
+            mps: Matrix Product State as a 'cirq.contrib.quimb.MPSState'.
+            bitstring: Bitstring |z⟩ as a binary string.
+        """
+        M_subset = []
+        for i, Ai in enumerate(mps.M):
+            qubit_index = mps.i_str(i)
+            # selecting the component with matching bitstring:
+            A_subset = Ai.isel({qubit_index: int(bitstring[i])})
```

Comment:
Just to check my understanding - this line is equivalent to the inner product of `Ai` with `bitstring[i]` (as a ket), correct?

Reply:
So their mps implementation is based off arXiv:2002.07730. If I recall correct, mps.M is a list of the tensors in equation (6) in that paper. The list has N tensors for N qubits, each has one set of indices ranging over the 2^N computational space and another which is related to the degree of entanglement.

The `i_str` function returns the tensor index for the ith qubit, and `isel` is analogous to list/array slicing given a tensor index label `qubit_index` and value `bitstring[i]`. So what this line is doing is building up a subset copy of the full tensor list.

Since we know the bitstring value of each qubit, this lets us reduce the 2^N indices to N as we consider a single bitstring rather than all possible N-bit bitstrings. 

As you mention in the next comment - it makes the final tensor contraction much faster and yields a scalar rather than TensorNetwork.